server/db/driver/pg/orders.go

```diff
@@ -0,0 +1,833 @@
+// error value is ErrUnknownOrder, and the type is
+// market/types.OrderStatusUnknown. The only recognized order types are market,
+// limit, and cancel.
+func (a *Archiver) Order(oid order.OrderID, base, quote uint32) (order.Order, types.OrderStatus, error) {
+	marketSchema, err := types.MarketName(base, quote)
+	if err != nil {
+		return nil, types.OrderStatusUnknown, err
+	}
+
+	// Since order type is unknown:
+	// - try to load from orders table, which includes market and limit orders
+	// - if found, coerce into the correct order type and return
+	// - if not found, try loading a cancel order with this oid
+	lo, status, err := loadLimitOrder(a.db, a.dbName, marketSchema, oid)
+	if err == ErrUnknownOrder {
+		// Try the cancel orders.
+		var co *order.CancelOrder
+		co, status, err = loadCancelOrder(a.db, a.dbName, marketSchema, oid)
+		if err != nil {
+			return nil, types.OrderStatusUnknown, err // includes ErrUnknownOrder
+		}
+		co.BaseAsset, co.QuoteAsset = base, quote
+		return co, pgToMarketStatus(status), err
+		// no other order types to try presently
+	}
+	if err != nil {
+		return nil, types.OrderStatusUnknown, err
+	}
+
+	lo.BaseAsset, lo.QuoteAsset = base, quote
+
+	// Since loadLimitOrder returns both market and limit orders in a
+	// *LimitOrder, identify the real type via the OrderType field. Extract the
+	// MarketOrder if necessary.
+	switch lo.OrderType {
+	case order.MarketOrderType:
+		return &lo.MarketOrder, pgToMarketStatus(status), nil
+	case order.LimitOrderType:
+		return lo, pgToMarketStatus(status), nil
+	}
+
+	return nil, types.OrderStatusUnknown,
+		fmt.Errorf("retrieved unsupported order type %d (%s)", lo.OrderType, lo.OrderType)
+}
+
+type pgOrderStatus uint16
+
+const (
+	orderStatusUnknown pgOrderStatus = iota
+	orderStatusEpoch
+	orderStatusBooked
+	orderStatusExecuted
+	orderStatusFailed // failed helps distinguish matched from unmatched executed cancel orders
+	orderStatusCanceled
+)
+
+func marketToPgStatus(status types.OrderStatus) pgOrderStatus {
+	switch status {
+	case types.OrderStatusEpoch:
+		return orderStatusEpoch
+	case types.OrderStatusBooked:
+		return orderStatusBooked
+	case types.OrderStatusExecuted:
+		return orderStatusExecuted
+	case types.OrderStatusCanceled:
+		return orderStatusCanceled
+	}
+	return orderStatusUnknown
+}
+
+func pgToMarketStatus(status pgOrderStatus) types.OrderStatus {
+	switch status {
+	case orderStatusEpoch:
+		return types.OrderStatusEpoch
+	case orderStatusBooked:
+		return types.OrderStatusBooked
+	case orderStatusExecuted, orderStatusFailed: // failed is executed as far as the market is concerned
+		return types.OrderStatusExecuted
+	case orderStatusCanceled:
+		return types.OrderStatusCanceled
+	}
+	return types.OrderStatusUnknown
+}
+
+func (status pgOrderStatus) String() string {
+	switch status {
+	case orderStatusFailed:
+		return "failed"
+	default:
+		return pgToMarketStatus(status).String()
+	}
+}
```

Comment:
`pgOrderStatus` serves two purposes:
1. Keep the definitions of status values used in the pg tables internal to this pg backend.  If the underlying integer value of a `market/types.OrderStatus` changes, it does not affect the value stored in the DB table.
2. Allows an additional status (failed) to help future queries distinguish matched from unmatched cancel orders that would otherwise both be in the executed state.